test/e2e/upgrade_management_cluster_test.go

```diff
@@ -35,307 +24,128 @@ func upgradeManagementCluster() {
-		clusterctl.UpgradeManagementClusterAndWait(ctx, clusterctl.UpgradeManagementClusterAndWaitInput{
-			ClusterctlConfigPath: clusterctlConfigPath,
-			ClusterProxy:         bootstrapClusterProxy,
-			Contract:             clusterv1.GroupVersion.Version,
-			LogFolder:            filepath.Join(artifactFolder, "clusters", bootstrapClusterProxy.GetName()),
-		}, e2eConfig.GetIntervals(specName, "wait-controllers")...)
-
-		Logf("The source cluster upgraded!")
+		}
+	}
 
-		By("Get the bootstrap cluster images after upgrading")
-		printImages(bootstrapClusterProxy)
+	// Create bmhs in the in the namespace that will host the new cluster
+	createBMH(clusterProxy, clusterNamespace, clusterName)
+	// This is exported here because we need this only when creating the workload
+	// if exported before it breaks creating the management because it uses v1beta1 templates
+	os.Setenv("PROVIDER_ID_FORMAT", "metal3://{{ ds.meta_data.uuid }}")
+}
 
-		By("UPGRADE MANAGEMENT CLUSTER PASSED!")
-	})
+// preInitFunc hook function that should be called from ClusterctlUpgradeSpec before init the management cluster
+// it installs certManager, BMO and Ironic and overrides the default IPs for the workload cluster.
+func preInitFunc(clusterProxy framework.ClusterProxy) {
+	installCertManager := func(clusterProxy framework.ClusterProxy) {
+		certManagerLink := fmt.Sprintf("https://github.com/cert-manager/cert-manager/releases/download/%s/cert-manager.yaml", e2eConfig.GetVariable("CERT_MANAGER_RELEASE"))
+		err := downloadFile("/tmp/certManager.yaml", certManagerLink)
+		Expect(err).To(BeNil(), "Unable to download certmanager manifest")
+		certManagerYaml, err := os.ReadFile("/tmp/certManager.yaml")
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(clusterProxy.Apply(ctx, certManagerYaml)).ShouldNot(HaveOccurred())
+	}
 
-	AfterEach(func() {
-		dumpSpecResourcesAndCleanup(ctx, specName, bootstrapClusterProxy, artifactFolder, namespace, e2eConfig.GetIntervals, clusterName, clusterctlLogFolder, skipCleanup)
-	})
+	// install certmanager
+	installCertManager(clusterProxy)
+	// Remove ironic
+	By("Remove Ironic containers from the source cluster")
+	ephemeralCluster := os.Getenv("EPHEMERAL_CLUSTER")
+	if ephemeralCluster == KIND {
+		removeIronicContainers()
+	} else {
+		removeIronicDeployment()
+	}
+	// install bmo
+	By("Install BMO")
+	installIronicBMO(clusterProxy, "false", "true")
+
+	// install ironic
+	By("Install Ironic in the target cluster")
+	installIronicBMO(clusterProxy, "true", "false")
+
+	// Export capi/capm3 versions
+	os.Setenv("CAPI_VERSION", "v1alpha4")
+	os.Setenv("CAPM3_VERSION", "v1alpha5")
+	// override default IPs for the workload cluster
+	os.Setenv("CLUSTER_APIENDPOINT_HOST", "192.168.111.250")
+	os.Setenv("BAREMETALV4_POOL_RANGE_START", "192.168.111.201")
+	os.Setenv("BAREMETALV4_POOL_RANGE_END", "192.168.111.240")
+	os.Setenv("PROVISIONING_POOL_RANGE_START", "172.22.0.201")
+	os.Setenv("PROVISIONING_POOL_RANGE_END", "172.22.0.240")
```

Comment:
same here, after all why do not use set them in the e2e conf file?

Reply:
this specific to the workload if we exported then before we will override the first management cluster